[DOCS] add details on version compatibility and remote gateway selection (#40056)
elastic/elasticsearch: docs/reference/modules/remote-clusters.asciidoc

endif::[]

Remote cluster connections work by configuring a remote cluster and connecting
only to a limited number of nodes in that remote cluster. Each remote cluster
is referenced by a name and a list of seed nodes. When a remote cluster is
registered, its cluster state is retrieved from one of the seed nodes and up
to three _gateway nodes_ are selected to be connected to as part of remote

> cluster requests. All the communication required between different clusters

Reviewer: > three _gateway nodes_ are selected to be connected to as part of remote cluster requests

It is unclear at this point in the page whether those three gateway nodes are used permanently or just for a particular request or whether the selection can change over time...?

Author: They are used for all CCR and CCS requests from then on. The selection may change if one or more of those nodes are shut down, in that case they will get replaced.

Reviewer: Thanks, I think we should add that piece of info here.

goes through the <<modules-transport,transport layer>>. Remote cluster
connections consist of uni-directional connections from the coordinating
node to the selected remote _gateway nodes_ only. By default each node in
a cluster that has remote clusters configured connects to one or more remote

> _gateway nodes_ and uses them to federate requests to the remote cluster.

Reviewer: > By default each node...

This sentence seems redundant to me.  Is there something important here that I'm missing?

Author: I tend to agree, I was about to remove it, but then I did not want to lose content. I will drop it.

[float]
[[gateway-nodes-selection]]
=== Gateway nodes selection

The _gateway nodes_ selection depends on the following criteria:

> - *version*: remote nodes need to be compatible with the cluster they are

Reviewer: ```suggestion
- *version*: Remote nodes must be compatible with the cluster they are
```

registered to. This is subject to the same rules as <<rolling-upgrades>>.
Any node can communicate with any other node on the same major version (e.g.
6.0 can talk to any 6.x node). Only nodes on the last minor version of a

> certain major version are able to communicate with nodes on the following major

Reviewer: ```suggestion
certain major version can communicate with nodes on the following major
```

version (e.g. 6.7 can communicate with 7.0, as well as any 7.x node, while
 6.6 or earlier cannot talk to any 7.x node). Note that version compatibility
 is symmetric, meaning that if 6.7 can communicate with 7.0, 7.0 can also
 communicate with 6.7. The matrix below summarizes compatibility as described
 above.

[cols="^,^,^,^,^,^"]
|====
| Compatibility | 5.0->5.5 | 5.6 | 6.0->6.6 | 6.7 | 7.x
| 5.0->5.5      |    Yes   | Yes |    No    | No  | No
| 5.6           |    Yes   | Yes |    Yes   | Yes | No
| 6.0->6.6      |    No    | Yes |    Yes   | Yes | No
| 6.7           |    No    | Yes |    Yes   | Yes | Yes
| 7.x           |    No    | No  |    No    | Yes | Yes
|====

> NOTE: When needing to connect incompatible clusters on subsequent major

Reviewer: ```suggestion
NOTE: If you want to connect incompatible clusters on subsequent major
```

> versions, it is possible to add to the cluster on the lower version one or

Reviewer: ```suggestion
versions, you can add one or
```

more <<coordinating-only-node,coordinating only nodes>> on the intermediate

> version that supports communicating with both ends. For instance, a 7.0

Reviewer: ```suggestion
version to the cluster on the lower version, since they support communicating with both ends. For instance, a 7.0
```

cluster can communicate with 6.6 or earlier nodes as long as at least one 6.7
_gateway node_ is present in their cluster. Beware that only features

> supported by the lower version can be used in that case.

Author: I will remove this note, I got feedback that it may be problematic for CCR and I also find it a bit confusing.

> - *role*: dedicated master nodes never get selected

Reviewer: ```suggestion
- *role*: Dedicated master nodes never get selected
```

> - *attributes*: you can tag which nodes should be selected

Reviewer: ```suggestion
- *attributes*: You can tag which nodes should be selected
```

(see <<remote-cluster-settings>>), though such tagged nodes still have

> to satisfy the two above requirements

Reviewer: ```suggestion
to satisfy the two above requirements.
```

You can use the <<cluster-remote-info, remote cluster info API>> to retrieve
information about the configured remote clusters, as well as the remote nodes

> that the node is connected to. This is particularly useful to verify which

Reviewer: > as well as the remote nodes that the node is connected to.

The API reference information (https://www.elastic.co/guide/en/elasticsearch/reference/master/cluster-remote-info.html) implies that "connected" is true/false.  Is that correct or does it list the gateway nodes?

Author: very good point. we do return a list of nodes but it's the list of seed nodes apparently rather than the selected gateway nodes, I need to adapt this then. Thanks for noticing!

> nodes get selected based on the criteria listed <<gateway-nodes-selection,above>>.

Reviewer: ```suggestion
nodes got selected according to the <<gateway-nodes-selection>>.
```
